lighthouse-cli/test/smokehouse/smokehouse.js

```diff
@@ -1,151 +1,226 @@
   }
+  if (failingTestCount) {
+    console.log(log.redify(`  ${failingTestCount} test(s) failing`));
+  }
+  console.log(''); // extra line break
 
-  /** @type {LH.Result} */
-  const lhr = JSON.parse(fs.readFileSync(outputPath, 'utf8'));
-  const artifacts = assetSaver.loadArtifacts(artifactsDirectory);
+  return {
+    id,
+    passingTestCount,
+    failingTestCount,
+  };
+}
+
+/** @param {string} str */
+function purpleify(str) {
+  return `${log.purple}${str}${log.reset}`;
+}
+
+/**
+ * Run Lighthouse in the selected runner. Returns `log`` for logging once
+ * all tests in a defn are complete.
+ * @param {{requestedUrl: string, configJson?: LH.Config.Json, expectation: Smokehouse.ExpectedRunnerResult, lighthouseRunner: LighthouseRunner, retries: number, isDebug?: boolean}} testOptions
+ * @return {Promise<{passed: number, failed: number, log: string}>}
+ */
+async function runSmokeTest(testOptions) {
+  const localConsole = new LocalConsole();
+  const {requestedUrl, configJson, expectation, lighthouseRunner, retries, isDebug} = testOptions;
+
+  // Rerun test until there's a passing result or retries are exhausted to prevent flakes.
+  let result;
+  let report;
+  for (let i = 0; i < retries; i++) {
+    if (i === 0) {
+      localConsole.log(`Doing a run of '${requestedUrl}'...`);
+    } else {
+      localConsole.log(`Retrying run (${i + 1} out of ${retries})...`);
+    }
+
+    // Run Lighthouse.
+    try {
+      result = await lighthouseRunner(requestedUrl, configJson, {isDebug});
+    } catch (e) {
+      logChildProcessError(localConsole, e);
+      continue; // Retry, if possible.
+    }
 
-  if (isDebug) {
-    console.log('LHR output available at: ', outputPath);
-    console.log('Artifacts avaiable in: ', artifactsDirectory);
-  } else {
-    fs.unlinkSync(outputPath);
-    rimraf.sync(artifactsDirectory);
+    // Assert result.
+    report = getAssertionReport(result, expectation, {isDebug});
+    if (report.failed) {
+      localConsole.log(`${report.failed} assertion(s) failed.`);
+      continue; // Retry, if possible.
+    }
+
+    break; // Passing result, no need to retry.
   }
 
-  // There should either be both an error exitCode and a lhr.runtimeError or neither.
-  if (Boolean(exitCode) !== Boolean(lhr.runtimeError)) {
-    const runtimeErrorCode = lhr.runtimeError && lhr.runtimeError.code;
-    console.error(`Lighthouse did not exit with an error correctly, exiting with ${exitCode} but with runtimeError '${runtimeErrorCode}'`); // eslint-disable-line max-len
-    process.exit(1);
+  // Write result log if we have one.
+  if (result) {
+    localConsole.write(result.log);
   }
 
+  // Without an assertion report, not much detail to share but a failure.
+  if (!report) {
+    return {
+      passed: 0,
+      failed: 1,
+      log: localConsole.getLog(),
+    };
+  }
+
+  localConsole.write(report.log);
   return {
-    lhr,
-    artifacts,
+    passed: report.passed,
```

Comment:
maybe a `success` boolean property too? set to `passingCount > 0 && failingCount === 0`

Reply:
> maybe a `success` boolean property too? set to `passingCount > 0 && failingCount === 0`

I also switched it to not caring about the total assertion count outside this method (reporting on tests passed/failed since `report()` already did the assertions within each test), so maybe it should drop the individual counts altogether for `success`?